src/js/tracing/reactnativenavigation.ts

```diff
@@ -0,0 +1,208 @@
+}
+
+const defaultOptions: ReactNativeNavigationOptions = {
+  routeChangeTimeoutMs: 1000,
+};
+
+interface ComponentEvent {
+  componentId: string;
+}
+
+type ComponentType =
+  | "Component"
+  | "TopBarTitle"
+  | "TopBarBackground"
+  | "TopBarButton";
+
+export interface ComponentWillAppearEvent extends ComponentEvent {
+  componentName: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  passProps?: Record<any, any>;
+  componentType: ComponentType;
+}
+
+export interface EventSubscription {
+  remove(): void;
+}
+
+export interface EventsRegistry {
+  registerComponentWillAppearListener(
+    callback: (event: ComponentWillAppearEvent) => void
+  ): EmitterSubscription;
+  registerCommandListener(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    callback: (name: string, params: any) => void
+  ): EventSubscription;
+}
+
+export interface NavigationDelegate {
+  events: () => EventsRegistry;
+}
+
+/**
+ * Instrumentation for React Native Navigation. See docs or sample app for usage.
+ *
+ * How this works:
+ * - `_onCommand` is called every time a commands happens and sets an IdleTransaction on the scope without any route context.
+ * - `_onComponentWillAppear` is then called AFTER the state change happens due to a dispatch and sets the route context onto the active transaction.
+ * - If `_onComponentWillAppear` isn't called within `STATE_CHANGE_TIMEOUT_DURATION` of the dispatch, then the transaction is not sampled and finished.
+ */
+export class ReactNativeNavigationInstrumentation extends RoutingInstrumentation {
+  public static instrumentationName: string = "react-native-navigation";
+
+  private _navigation: NavigationDelegate;
+  private _options: ReactNativeNavigationOptions;
+
+  private _prevComponentEvent: ComponentWillAppearEvent | null = null;
+
+  private _latestTransaction?: TransactionType;
+  private _recentComponentIds: string[] = [];
+  private _stateChangeTimeout?: number | undefined;
+
+  public constructor(
+    /** The react native navigation `NavigationDelegate`. This is usually the import named `Navigation`. */
+    navigation: NavigationDelegate,
+    options: Partial<ReactNativeNavigationOptions> = {}
+  ) {
+    super();
+
+    this._navigation = navigation;
+
+    this._options = {
+      ...defaultOptions,
+      ...options,
+    };
+  }
+
+  /**
+   * Registers the event listeners for React Native Navigation
+   */
+  public registerRoutingInstrumentation(
+    listener: TransactionCreator,
+    beforeNavigate: BeforeNavigate
+  ): void {
+    super.registerRoutingInstrumentation(listener, beforeNavigate);
+
+    this._navigation
+      .events()
+      .registerCommandListener(this._onCommand.bind(this));
+
+    this._navigation
+      .events()
+      .registerComponentWillAppearListener(
+        this._onComponentWillAppear.bind(this)
+      );
+  }
+
+  /**
+   * To be called when a navigation command is dispatched
+   */
+  private _onCommand(): void {
```

Comment:
Ahh this feels like a https://en.wikipedia.org/wiki/Finite-state_machine, but it's probably too much effort to refactor the logic into a state machine.